contracts/hooks/FullRange.sol

```diff
@@ -0,0 +1,492 @@
+        internal
+        returns (BalanceDelta delta)
+    {
+        // msg.sender is the test contract (aka whoever called addLiquidity/removeLiquidity)
+
+        delta = abi.decode(poolManager.lock(abi.encode(CallbackData(msg.sender, key, params, false))), (BalanceDelta));
+
+        uint256 ethBalance = address(this).balance;
+        if (ethBalance > 0) {
+            CurrencyLibrary.NATIVE.transfer(msg.sender, ethBalance);
+        }
+    }
+
+    function hookModifyPosition(IPoolManager.PoolKey memory key, IPoolManager.ModifyPositionParams memory params)
+        internal
+        returns (BalanceDelta delta)
+    {
+        // msg.sender is the test contract (aka whoever called addLiquidity/removeLiquidity)
+
+        delta = abi.decode(poolManager.lock(abi.encode(CallbackData(address(this), key, params, true))), (BalanceDelta));
+
+        uint256 ethBalance = address(this).balance;
+        if (ethBalance > 0) {
+            CurrencyLibrary.NATIVE.transfer(msg.sender, ethBalance);
+        }
+    }
+
+    function lockAcquired(uint256, bytes calldata rawData) external override returns (bytes memory) {
+        require(msg.sender == address(poolManager));
+
+        CallbackData memory data = abi.decode(rawData, (CallbackData));
+
+        BalanceDelta delta = poolManager.modifyPosition(data.key, data.params);
+
+        HookPosition storage position = poolToHookPosition[data.key.toId()];
+
+        // this does all the transfers
+        if (delta.amount0() > 0) {
+            if (data.key.currency0.isNative()) {
+                poolManager.settle{value: uint128(delta.amount0())}(data.key.currency0);
+            } else {
+                IERC20Minimal(Currency.unwrap(data.key.currency0)).transferFrom(
+                    data.sender, address(poolManager), uint128(delta.amount0())
+                );
+                poolManager.settle(data.key.currency0);
+            }
+        } else {
+            if (data.rebalance) {
+                // retrieve all fees -
+                uint256 feeGrowthInside0LastX128 =
+                    poolManager.getPosition(data.key.toId(), address(this), MIN_TICK, MAX_TICK).feeGrowthInside0LastX128;
+
+                position.tokensOwed0 += uint128(-delta.amount0())
+                    + uint128(
+                        FullMath.mulDiv(
+                            feeGrowthInside0LastX128 - position.feeGrowthInside0LastX128,
+                            position.liquidity,
+                            FixedPoint128.Q128
+                        )
+                    );
+
+                uint256 balBefore = balanceOf(data.key.currency0, data.sender);
+                poolManager.take(data.key.currency0, data.sender, position.tokensOwed0);
+                uint256 balAfter = balanceOf(data.key.currency0, data.sender);
+                require(balAfter - balBefore == position.tokensOwed0);
```

Comment:
check seems redundant since take will most definitely return the the amount it's given? (unless it reverts)